File: engine/argot/scoring/evidence/call_receiver.py
"""Collector for :class:`CallReceiverEvidence` payloads.

Cluster-scoped: the call-receiver scorer groups files by callee-bag
similarity, and a callee that's typical in one cluster (a logger in a
service module) can be unfamiliar in another (a CLI entry point). The
``common here:`` slice and the rarity denominator therefore both come
from the hunk file's MinHash cluster, not the repo as a whole.

Singleton / unknown clusters fall back to a repo-empty framing rather
than printing a wrong cluster's data — see the docstring on the
collector for the exact contract.
"""

from __future__ import annotations

from argot.scoring.evidence.types import CallReceiverEvidence, RarityStat
from argot.scoring.evidence.types import EvidenceCorpus as _EvidenceCorpus

_COMMON_HERE_LIMIT = 10
_CALLEE_NOUN = "callees"
_CLUSTER_WHERE = "this cluster"
_REPO_WHERE = "repo"


def collect_call_receiver_evidence(
    *,
    unattested_callees: list[str],
    cluster_id: int | None,
    evidence_corpus: _EvidenceCorpus,
) -> CallReceiverEvidence:
    """Build :class:`CallReceiverEvidence` scoped to the hunk's cluster.

    ``cluster_id`` is the cluster the call-receiver scorer assigned to the
    hunk's file (``None`` for files that fell through the static lookup
    *and* the Jaccard fallback — singleton clusters or files with empty
    callee bags). When ``None``, ``common_here`` is empty and the rarity
    denominator is 0; the formatter then prints "never seen in repo"
    instead of "0 of N in this cluster" — honest about the missing
    cluster context rather than printing a misleading denominator.
    """
    if cluster_id is not None and cluster_id in evidence_corpus.callees_by_cluster:
        common = list(evidence_corpus.callees_by_cluster[cluster_id][:_COMMON_HERE_LIMIT])
        denom = evidence_corpus.totals.callees_attested_by_cluster.get(cluster_id, 0)
        where = _CLUSTER_WHERE
    else:
        common = []
        denom = 0
        where = _REPO_WHERE
    return CallReceiverEvidence(
        unfamiliar_callees=list(unattested_callees),
        rarity=RarityStat(
            flagged_count=0,
            attested_total=denom,
            noun=_CALLEE_NOUN,
            where=where,
        ),
        common_here=common,
    )


__all__ = ["collect_call_receiver_evidence"]

File: crates/argot-rules-voice/src/scoring/evidence/call_receiver.rs
//! Collector for [`CallReceiverEvidence`] payloads.
//!
//! Port of `engine/argot/scoring/evidence/call_receiver.py`. Cluster-scoped:
//! the `common here:` slice and the rarity denominator come from the hunk
//! file's MinHash cluster. Singleton / unknown clusters fall back to a
//! repo-empty framing rather than printing a wrong cluster's data.

use super::types::{CallReceiverEvidence, CommonEntry, EvidenceCorpus, RarityStat};

const COMMON_HERE_LIMIT: usize = 10;
const CALLEE_NOUN: &str = "callees";
const CLUSTER_WHERE: &str = "this cluster";
const REPO_WHERE: &str = "repo";

/// Build [`CallReceiverEvidence`] scoped to the hunk's cluster. When
/// `cluster_id` is `None` (or absent from the corpus), `common_here` is empty
/// and the denominator is 0, so the formatter prints "never seen in repo"
/// rather than a misleading cluster denominator.
pub fn collect_call_receiver_evidence(
    unattested_callees: Vec<String>,
    cluster_id: Option<usize>,
    evidence_corpus: &EvidenceCorpus,
) -> CallReceiverEvidence {
    let (common_here, denom, where_) = match cluster_id {
        Some(cid) if evidence_corpus.callees_by_cluster.contains_key(&cid) => {
            let common: Vec<CommonEntry> = evidence_corpus.callees_by_cluster[&cid]
                .iter()
                .take(COMMON_HERE_LIMIT)
                .cloned()
                .collect();
            let denom = evidence_corpus
                .totals
                .callees_attested_by_cluster
                .get(&cid)
                .copied()
                .unwrap_or(0);
            (common, denom, CLUSTER_WHERE)
        }
        _ => (Vec::new(), 0, REPO_WHERE),
    };
    CallReceiverEvidence {
        unfamiliar_callees: unattested_callees,
        rarity: RarityStat {
            flagged_count: 0,
            attested_total: denom,
            noun: CALLEE_NOUN.to_string(),
            where_: where_.to_string(),
        },
        common_here,
    }
}

#[cfg(test)]
mod tests;
